commit f36b0d9d4053e167f59b89286cc9378e8e1563f7
Author: Tim Serong <tserong@suse.com>
Date:   Fri May 12 22:24:19 2017 +1000

    mgr: use new Gil class in place of PyGILState_*() API
    
    Prep work for loading modules in separate python sub-interpreters.
    According to the Python C API docs, mixing the PyGILState_*() API
    with sub-interpreters is unsupported, so I've replaced these calls
    with a Gil class, which will acquire and release the GIL against
    a specific python thread state.
    
    Note the manual python thread state creation in MgrPyModule::serve().
    The PyGILState_*() APIs would have done that for us, but we're not
    using them anymore, so have to do it by hand (see
    https://docs.python.org/2/c-api/init.html#non-python-created-threads
    for some description of this).
    
    Signed-off-by: Tim Serong <tserong@suse.com>

diff --git a/src/mgr/PyModules.cc b/src/mgr/PyModules.cc
index 10f4ce4787d..8cc3f674c0b 100644
--- a/src/mgr/PyModules.cc
+++ b/src/mgr/PyModules.cc
@@ -13,6 +13,7 @@
 
 // Include this first to get python headers earlier
 #include "PyState.h"
+#include "Gil.h"
 
 #include <boost/tokenizer.hpp>
 #include "common/errno.h"
@@ -403,16 +404,21 @@ int PyModules::init()
     PyEval_InitThreads();
   }
 
+  // Drop the GIL and remember the main thread state (current
+  // thread state becomes NULL)
+  pMainThreadState = PyEval_SaveThread();
+
   // Load python code
   boost::tokenizer<> tok(g_conf->mgr_modules);
   for(const auto& module_name : tok) {
     dout(1) << "Loading python module '" << module_name << "'" << dendl;
-    auto mod = std::unique_ptr<MgrPyModule>(new MgrPyModule(module_name));
+    auto mod = std::unique_ptr<MgrPyModule>(new MgrPyModule(module_name, pMainThreadState));
     int r = mod->load();
     if (r != 0) {
+      // Don't use handle_pyerror() here; we don't have the GIL
+      // or a thread state (this is deliberate).
       derr << "Error loading module '" << module_name << "': "
         << cpp_strerror(r) << dendl;
-      derr << handle_pyerror() << dendl;
       // Don't drop out here, load the other modules
     } else {
       // Success!
@@ -420,9 +426,6 @@ int PyModules::init()
     }
   } 
 
-  // Drop the GIL
-  PyEval_SaveThread();
-  
   return 0;
 }
 
@@ -438,15 +441,12 @@ public:
 
   void *entry() override
   {
-    PyGILState_STATE gstate;
-    gstate = PyGILState_Ensure();
     running = true;
 
+    // No need to acquire the GIL here; the module does it.
     dout(4) << "Entering thread for " << mod->get_name() << dendl;
     mod->serve();
 
-    PyGILState_Release(gstate);
-
     running = false;
     return nullptr;
   }
@@ -495,7 +495,7 @@ void PyModules::shutdown()
 
   modules.clear();
 
-  PyGILState_Ensure();
+  PyEval_RestoreThread(pMainThreadState);
   Py_Finalize();
 
   // nobody needs me anymore.
diff --git a/src/mgr/PyModules.h b/src/mgr/PyModules.h
index c68366f2a17..4c858747306 100644
--- a/src/mgr/PyModules.h
+++ b/src/mgr/PyModules.h
@@ -43,6 +43,8 @@ class PyModules
 
   std::string get_site_packages();
 
+  PyThreadState *pMainThreadState = nullptr;
+
 public:
   static std::string config_prefix;
 
